OIDC example: restrict role to specific branch (#7902)
circleci/circleci-docs: jekyll/_cci2/openid-connect-tokens.adoc

Consult your cloud service's documentation for steps to add an identity provider. For example, AWS's https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html[Creating OpenID Connect (OIDC) identity providers], or Google Cloud Platform's https://cloud.google.com/iam/docs/configuring-workload-identity-federation#oidc[Configuring workload identity federation].

> The https://openid.net/specs/openid-connect-core-1_0.html#Terminology[OpenID Provider] is unique to your organization. The URL is `https://oidc.circleci.com/org/<organization_id>`, where `organization_id` is the organization ID (a universally unique identifier) that represents your organization. You can find your CircleCI organization ID by navigating to **Organization Settings > Overview** in the https://app.circleci.com/[CircleCI web app].

Author: @rosieyohannan I'm not sure why, but in my dev preview, a `;` gets appended to the URL i.e. `https://oidc.circleci.com/org/<organization_id>;`

Reviewer: Strange! in your local build you mean?

Author: @rosieyohannan Yes. Any chance you could check my branch locally?

Visit the https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html[Creating OpenID Connect (OIDC) identity providers] page of the AWS docs and follow the instructions. You will need your CircleCI Organization ID, which you can find by navigating to **Organization Settings > Overview** on the https://app.circleci.com/[CircleCI web app]. Copy your Organization ID for the next step.

> When asked for the **Provider URL**, enter: `\https://oidc.circleci.com/org/<organization_id>`, where `organization_id` is the ID of your CircleCI organization. Provide the same CircleCI Organization ID for the **Audience**.

Author: @rosieyohannan When we indicate that part of a string needs to be replaced with a value, should we include the `< >` , e.g. "...where `<organization_id>` is the ID of your CircleCI org"?

Reviewer: I think the way you have it there without `< >` in the text is right 👍

> To do this, edit your IAM role's trust policy so that only an OIDC token from your chosen project can assume that role. The trust policy determines under what conditions the role can be assumed.

Reviewer: What do you think about adding another heading level here, so like `===== Restrict by project` and `===== Restrict by branch` or something?

Reviewer: Also we're using different methods of indicating items the reader needs to substitute in code snippets on this page.  Would be good to change to lowercase and within `< >` for all on this page. Some instances of this are outside the scope of your changes here actually so I will make another ticket to avoid holding you up :-D.

> . Go to an individual project's page on https://app.circleci.com/[CircleCI web app] and navigate to **Project Settings > Overview** to find your Project ID.

Reviewer: ```suggestion
. Go to an individual project's page in the https://app.circleci.com/[CircleCI web app] and navigate to **Project Settings > Overview** to find your Project ID.
```

. Add the following condition to your role's trust policy, so that only jobs in your chosen project can assume that role. Enter your Organization ID for `ORGANIZATION_ID` and your Project ID for `PROJECT_ID`.
+
```yaml
"StringLike": {
  "oidc.circleci.com/org/ORGANIZATION_ID:sub": "org/ORGANIZATION_ID/project/PROJECT_ID/user/*"
}
```
+
This uses https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_condition_operators.html#Conditions_String[StringLike] to match the sub claim of CircleCI's OIDC token in your chosen project. Now, jobs in your other projects cannot assume this role.

> You can also restrict access to specific branches. The following is an example that restricts the `AssumeRoleWithWebIdentity` action to any project pipelines running only on the `main` branch in the `my-org` GitHub organization and `6b376885-8f20-45b2-ba2f-c27c0919bc6d` CircleCI organization. 

Reviewer: ```suggestion
You can also restrict access to specific branches. The following is an example that restricts the `AssumeRoleWithWebIdentity` action to pipelines running on the `main` branch in the `my-org` GitHub organization, and the CircleCI organization with the ID: `6b376885-8f20-45b2-ba2f-c27c0919bc6d`. 
```
